Port StrictMath.floor and ceil from fdlibm

tanishiking opened the pull request:
- https://www.netlib.org/fdlibm/s_ceil.c
- https://www.netlib.org/fdlibm/s_floor.c

And add corner case tests.

tanishiking committed:
StrictMath.floor ported from fdlibm

https://www.netlib.org/fdlibm/s_floor.c

And add some whitebox tests

tanishiking committed:
StrictMath.ceil ported from fdlibm

https://www.netlib.org/fdlibm/s_ceil.c

tanishiking committed:
Enable tests that uses Math.floor

sjrd commented on javalib/src/main/scala/java/lang/StrictMath.scala:
```diff
@@ -0,0 +1,187 @@
+object StrictMath {
+
+  // Ported from https://www.netlib.org/fdlibm/s_floor.c
+  def floor(x: scala.Double): scala.Double = {
```
`floor` and `ceil` are among the subset of methods for which `Math.floor` is exactly specified. Therefore, `StrictMath` and `Math` do the same thing. I suggest we keep this implementation inside `Math` for those methods, in order to reduce code size.

tanishiking replied:
Ah I see, indeed `StrictMath.floor` and `Math.floor` have exactly the same description.
Only the methods like trigonometric/exponents/logarithmic whose output is strictly defined in `StrictMath` should be in that package :+1:

sjrd commented on javalib/src/main/scala/java/lang/StrictMath.scala:
```diff
@@ -0,0 +1,187 @@
+      Double.longBitsToDouble(bits)
+    }
+
+    if (exponent < 20) { // Case 1: decimal point is withing the high word.
```
TBH, this seems like an antiquated way of writing this algorithm. fdlibm was written in a time where 64-bit integers were not native, and it shows. Working with the full `Long` value would be a lot easier to understand, IMO. (not to mention efficient, but that's not important for those slow-path methods)

There is value in directly following fdlibm to guarantee the exact same results. But for an exactly-specified method like `floor`, that argument is not so important. A modern implementation that is more understandable is probably wiser.

tanishiking replied:
Ah true, I tried to port the implementation as is, but yeah splitting to low and high seems only for using 32-bit integer operations for performance (at the time when 64-bit integers weren't native).
Just using 64-bit integer makes it much simpler :+1:

sjrd commented on javalib/src/main/scala/java/lang/StrictMath.scala:
```diff
@@ -0,0 +1,187 @@
+
+  // Ported from https://www.netlib.org/fdlibm/s_floor.c
+  def floor(x: scala.Double): scala.Double = {
+    val doubleBits = Double.doubleToLongBits(x)
```
```suggestion
    val doubleBits = Double.doubleToRawLongBits(x)
```
No need to normalize NaNs. By default we use the raw variant when it makes no difference.

sjrd commented on javalib/src/main/scala/java/lang/StrictMath.scala:
```diff
@@ -0,0 +1,187 @@
+              (highWord + 1, lowWord)
+            } else {
+              val newLowWord = lowWord + (1 << (52 - exponent)) // + LSB of integer part
+              if (Integer.compareUnsigned(newLowWord, lowWord) < 0) { // carry
```
```suggestion
              if (Integer.unsigned_<(newLowWord, lowWord)) { // carry
```
(but you won't need that with a `Long`-based implementation at all)

tanishiking replied:
Yup, we no longer need carry with a Long-based implementation.

sjrd commented on javalib/src/main/scala/java/lang/StrictMath.scala:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  // Ported from https://www.netlib.org/fdlibm/s_ceil.c
+  def ceil(x: scala.Double): scala.Double = {
```
Since these methods are slow paths anyway (only used when the optimizer is turned off), consider delegating to `floor` instead of implementing things twice. The spec is such that `ceil(x)` is the same as `-floor(-x)`.

sjrd commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/StrictMathTest.scala:
```diff
@@ -0,0 +1,98 @@
+    assertExactEquals(0.0, StrictMath.floor(0.5))
+    assertExactEquals(-1.0, StrictMath.floor(-0.5))
+
+    // Special values
```
For consistency with other tests, put those first, and also include `+0.0` and `-0.0`.

tanishiking committed:
Do not split into two 32-bits in StrictMath.floor

tanishiking committed:
Keep implementing floor and ceil in Math instead of StrictMath

They are among the subset of methods for which Math.floor is exactly specified.

sjrd commented on javalib/src/main/scala/java/lang/Math.scala:
```diff
@@ -126,8 +138,52 @@ object Math {
+  @inline def floor(a: scala.Double): scala.Double = {
+    linkTimeIf(LinkingInfo.targetPureWasm) {
+      // Ported from https://www.netlib.org/fdlibm/s_floor.c
+      var bits = Double.doubleToRawLongBits(a)
```
For consistency (and arguably readability), put the pure-Wasm implementation in a separate `private def`?

sjrd commented on javalib/src/main/scala/java/lang/Math.scala:
```diff
@@ -126,8 +138,52 @@ object Math {
+         * "if (huge + x > 0.0)" (huge = 1.0e300),
+         * which was used to raise the IEEE 754 inexact flag.
+         * In Java, there is no standard way to access these flags,
+         * so this check has no practical effect(?) and commented out.
```
```suggestion
         * so this check has no practical effect and was removed.
```
?
